selene-lib/src/rules/mismatched_arg_count.rs

```diff
@@ -0,0 +1,425 @@
+                        passed_argument_count += 1;
+                        if expression.has_side_effects() {
+                            return PassedArgumentCount::Variable(passed_argument_count);
+                        }
+                    }
+                }
+            }
+
+            PassedArgumentCount::Fixed(passed_argument_count)
+        }
+        ast::FunctionArgs::String(_) => PassedArgumentCount::Fixed(1),
+        ast::FunctionArgs::TableConstructor(_) => PassedArgumentCount::Fixed(1),
+    }
+}
+
+/// A visitor used to map a variable to the necessary number of parameters required
+struct MapFunctionDefinitionVisitor<'a> {
+    scope_manager: &'a ScopeManager,
+    definitions: &'a mut HashMap<Id<Variable>, ParameterCount>,
+    /// Blacklisted variables are ones we will ignore completely - this may be because its a reassigned global variable
+    /// so we cannot determine which function body/parameter count matches.
+    blacklisted_variables: HashSet<Id<Variable>>,
+}
+impl Visitor<'_> for MapFunctionDefinitionVisitor<'_> {
+    fn visit_local_function(&mut self, function: &ast::LocalFunction<'_>) {
+        let identifier = range(function.name());
+        let variable = self
+            .scope_manager
+            .variables
+            .iter()
+            .find(|v| v.1.identifiers.contains(&identifier));
+
+        if let Some((id, _)) = variable {
+            self.definitions
+                .insert(id, ParameterCount::from_function_body(function.func_body()));
+        }
+    }
+
+    fn visit_function_declaration(&mut self, function: &ast::FunctionDeclaration<'_>) {
+        let identifier = range(function.name());
+        if let Some((_, reference)) = self
+            .scope_manager
+            .references
+            .iter()
+            .find(|r| r.1.identifier == identifier)
+        {
+            if let Some(variable) = reference.resolved {
+                // Ignore blacklisted variables
+                // Check to see we aren't reassigning an already assigned variable
+                // If we are - bail out of this and blacklist it, as we don't know which function definition is correct
+                if self.blacklisted_variables.contains(&variable) {
+                    return;
+                } else if self.definitions.contains_key(&variable) {
+                    self.definitions.remove(&variable);
+                    self.blacklisted_variables.insert(variable);
+                } else {
+                    self.definitions.insert(
+                        variable,
+                        ParameterCount::from_function_body(function.body()),
+                    );
+                }
+            }
+        }
+    }
+
+    fn visit_local_assignment(&mut self, local_assignment: &ast::LocalAssignment) {
+        let mut expressions = local_assignment.expr_list().iter();
+
+        for name_token in local_assignment.name_list() {
+            let expression = expressions.next();
```

Comment:
Use `zip`.

Reply:
I did try using `zip` here, with something like `let assignment_expressions = assignment.var_list().iter().zip(assignment.expr_list());`, but I had some awkward lifetime errors, so I just used this aswell (which I got from ScopeManager).
I'll try this again though, and see if I can find what went wrong

Reply:
Nevermind, seems to be working fine now - must have done something wrong last time!